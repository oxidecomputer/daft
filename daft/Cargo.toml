[package]
name = "daft"
version = "0.1.0"
description = "Structural diffs of Rust data structures"
edition.workspace = true
repository.workspace = true
rust-version.workspace = true
license.workspace = true

[lints]
workspace = true

[dependencies]
daft-derive.workspace = true
newtype-uuid = { workspace = true, optional = true }
oxnet = { workspace = true, optional = true }
paste.workspace = true
uuid = { workspace = true, optional = true, features = ["v4"] }

[features]
newtype-uuid1 = ["dep:newtype-uuid"]
oxnet01 = ["dep:oxnet"]
uuid1 = ["dep:uuid"]
