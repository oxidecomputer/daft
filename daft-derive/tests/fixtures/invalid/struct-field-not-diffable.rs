use daft::Diffable;

struct NonDiffable {}

#[derive(Diffable)]
struct MyStruct {
    a: i32,
    b: NonDiffable,
}

fn main() {
    // MyStruct should still exist, even though the Diffable impl has errors.
    let _ = MyStruct { a: 0, b: NonDiffable {} };
}
