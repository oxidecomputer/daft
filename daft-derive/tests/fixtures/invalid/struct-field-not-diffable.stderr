error[E0277]: the trait bound `NonDiffable: Diffable` is not satisfied
 --> tests/fixtures/invalid/struct-field-not-diffable.rs:8:5
  |
8 |     b: NonDiffable,
  |     ^ the trait `Diffable` is not implemented for `NonDiffable`
  |
  = help: the following other types implement trait `Diffable`:
            &'a T
            ()
            (A, B)
            (A, B, C)
            (A, B, C, D)
            (A, B, C, D, E)
            (A, B, C, D, E, F)
            (A, B, C, D, E, F, G)
          and $N others

error[E0277]: the trait bound `NonDiffable: Diffable` is not satisfied
 --> tests/fixtures/invalid/struct-field-not-diffable.rs:5:10
  |
5 | #[derive(Diffable)]
  |          ^^^^^^^^ the trait `Diffable` is not implemented for `NonDiffable`
  |
  = help: the following other types implement trait `Diffable`:
            &'a T
            ()
            (A, B)
            (A, B, C)
            (A, B, C, D)
            (A, B, C, D, E)
            (A, B, C, D, E, F)
            (A, B, C, D, E, F, G)
          and $N others
  = note: this error originates in the derive macro `Diffable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NonDiffable: Diffable` is not satisfied in `MyStructDiff<'__daft>`
 --> tests/fixtures/invalid/struct-field-not-diffable.rs:5:10
  |
5 | #[derive(Diffable)]
  |          ^^^^^^^^ within `MyStructDiff<'__daft>`, the trait `Diffable` is not implemented for `NonDiffable`
  |
  = help: the following other types implement trait `Diffable`:
            &'a T
            ()
            (A, B)
            (A, B, C)
            (A, B, C, D)
            (A, B, C, D, E)
            (A, B, C, D, E, F)
            (A, B, C, D, E, F, G)
          and $N others
note: required because it appears within the type `MyStructDiff<'__daft>`
 --> tests/fixtures/invalid/struct-field-not-diffable.rs:5:10
  |
5 | #[derive(Diffable)]
  |          ^^^^^^^^
note: required by a bound in `daft::Diffable::Diff`
 --> $WORKSPACE/daft/src/diffable.rs
  |
  | /     type Diff<'daft>
  | |     where
  | |         Self: 'daft;
  | |____________________^ required by this bound in `Diffable::Diff`
  = note: this error originates in the derive macro `Diffable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NonDiffable: Diffable` is not satisfied in `MyStructDiff<'__daft>`
 --> tests/fixtures/invalid/struct-field-not-diffable.rs:5:10
  |
5 | #[derive(Diffable)]
  |          ^^^^^^^^ within `MyStructDiff<'__daft>`, the trait `Diffable` is not implemented for `NonDiffable`
  |
  = help: the following other types implement trait `Diffable`:
            &'a T
            ()
            (A, B)
            (A, B, C)
            (A, B, C, D)
            (A, B, C, D, E)
            (A, B, C, D, E, F)
            (A, B, C, D, E, F, G)
          and $N others
note: required because it appears within the type `MyStructDiff<'__daft>`
 --> tests/fixtures/invalid/struct-field-not-diffable.rs:5:10
  |
5 | #[derive(Diffable)]
  |          ^^^^^^^^
  = note: the return type of a function must have a statically known size
  = note: this error originates in the derive macro `Diffable` (in Nightly builds, run with -Z macro-backtrace for more info)
